backend/api/quivr_api/modules/sync/utils/sync.py

```diff
@@ -636,3 +697,316 @@ def download_file(
+        children_blocks = await self.notion.blocks.children.list(page_id)
+        children_blocks = children_blocks["results"]
+        n_child_page = len(
+            [
+                block
+                for block in children_blocks
+                if block["type"] in ["child_page", "child_database"]
+            ]
+        )
+        if n_child_page > 0:
+            return True
+        return False
+
+    async def get_icon(self, page_id):
+        try:
+            page = await self.notion.pages.retrieve(page_id)
+            if page["icon"] is None:
+                return {"emoji": None}
+            return page["icon"]
+        except Exception as e:
+            logger.error("Error retrieving Notion file with ID %s: %s", page_id, e)
+            return {"emoji": "📊"}
+
+    async def fetch_icons_and_folder_statuses(self, blocks):
+        icon_tasks = {
+            block["id"]: asyncio.create_task(self.get_icon(block["id"]))
+            for block in blocks
+        }
+        folder_tasks = {
+            block["id"]: asyncio.create_task(self.is_folder(block["id"]))
+            for block in blocks
+        }
+
+        # Wait for all icon tasks to complete
+        icons = await asyncio.gather(*icon_tasks.values())
+        # Wait for all folder status tasks to complete
+        folder_statuses = await asyncio.gather(*folder_tasks.values())
+
+        # Combine results
+        result = {}
+        for block_id, icon, is_folder in zip(icon_tasks.keys(), icons, folder_statuses):
+            result[block_id] = {"icon": icon["emoji"], "is_folder": is_folder}
+
+        return result
+
+    async def fetch_pages(self, page_id, recursive) -> List[SyncFile]:
+        pages = []
+        blocks = await self.notion.blocks.children.list(page_id)
+        blocks = blocks["results"]  # type: ignore
+        icon_is_folder = await self.fetch_icons_and_folder_statuses(blocks)
+
+        for block in blocks:
+            block_type = block["type"]
+            # is_folder = True
+            if block_type in {"child_database", "child_page"}:
+                # if block_type is child database mark it as unclickable
+                page_info = SyncFile(
+                    name=f'{block[block_type]["title"]}.md',
+                    id=block["id"],
+                    is_folder=icon_is_folder[block["id"]]["is_folder"],
+                    last_modified=block["last_edited_time"],
+                    mime_type="md" if block_type == "child_page" else "db",
+                    web_view_link=f"https://www.notion.so/{block['id'].replace('-', '')}",
+                    icon=(
+                        icon_is_folder[block["id"]]["icon"]
+                        if icon_is_folder[block["id"]]["icon"]
+                        else None
+                    ),
+                )
+                pages.append(page_info)
+
+                if recursive:
+                    sub_pages = await self.fetch_pages(block["id"], recursive)
```

Comment:
This could be parallelized using async queues. Basically spawn a set of tasks that polls on the queue. If the `recursive=True` the page will put the `block["id"]` in the queue.

We will then wait on all tasks to be done with the processing (queue empty)